docs/discussion/Ratify v2 design scope.md

```diff
@@ -0,0 +1,154 @@
+      - [Configuration Refactoring](#configuration-refactoring)
+      - [Reorganize the Codebase](#reorganize-the-codebase)
+    - [Deprecations](#deprecations)
+      - [CertificateStore](#certificatestore)
+      - [Legacy Cosign Verifier](#legacy-cosign-verifier)
+      - [License Checker Verifier](#license-checker-verifier)
+      - [`Name` and `Type` fields in verifierReport/configuration](#name-and-type-fields-in-verifierreportconfiguration)
+    - [Documentation Update](#documentation-update)
+    - [Unit Test Improvement](#unit-test-improvement)
+
+## Background
+Ratify has reached its first major release in over a year and has been widely adopted by users at scale. However, there are known limitations in the current design and implementation of Ratify v1 make it challenging to extend functionality. This document outlines these known limitations and proposes solutions for a more lightweight and extensible framework.
+
+## Goal
+
+The goal of the v2 design is to improve the current design and implementation to better meet the requirements of the customers.
+
+1. Streamline our offerings by phasing out outdated features, ensuring customers have a clear view of all actively supported features.
+2. Refactor the code to make it more maintainable and extensible for new features.
+3. Improve the performance of the system.
+4. Improve the usability of the system.
+5. Reduce the dependencies of the system.
+
+## Scope
+
+Overall, the scope of the v2 design includes the following aspects: system refactoring, deprecations, document update and unit test improvement. We'll discuss each aspect in detail.
+
+Notes: The priority of each task is scored from 1 to 5, with 1 being the highest priority.
+
+### Refactoring
+
+#### Plugin Framework Refactoring
+**Current Limitation**:
+1. External and internal verifiers are not sharing the same in-memory cache. Each external verifier will have a separate ReferrerStore while created. This could lead to data inconsistency and performance issues.
+2. The OCI store behind the ORAS store is not process-safe. This could lead to race conditions when multiple external/interval verifiers are running in parallel. We already have a tracking [issue](https://github.com/ratify-project/ratify/issues/1110).
+3. The built-in plugins are tightly coupled with the Ratify core. Plugin version upgrade requires a new Ratify release. And built-in plugins also add dependencies to the Ratify core which makes it hard to maintain.
+
+**Proposed Solution**:
+1. We need to refactor the whole plugin framework to make it more lightweight and concurrent-safe. Specifically, we have 2 options:
+   1. Use a plugin framework like [HashiCorp go-plugin](https://github.com/hashicorp/go-plugin) to make Ratify main process communicate with external verifiers through GRPC. And we can have Ratify main process to manage the cache and OCI store while external verifiers will query/set cache through GRPC call.
```

Comment:
curious if we use grpc , will each instance of plugin be a long living service?

Reply:
yes, I'm a bit concerned on the performance impact if everything is over gRPC. Probably need to do some benchmark once I have the prototype or just seek for another solution.